src/js/jsx/tools/ArtboardAdderOverlay.jsx

```diff
@@ -0,0 +1,305 @@
+                modalState: modalState,
+                vectorMaskMode: vectorMaskMode
+            };
+        },
+
+        shouldComponentUpdate: function (nextProps, nextState) {
+            // We check for the cheaper flags before plucking layers from document models
+            if (this.state.vectorMaskMode !== nextState.vectorMaskMode ||
+                this.state.modalState !== nextState.modalState ||
+                this.props.transformString !== nextProps.transformString) {
+                return true;
+            }
+
+            var lastLayers = this.state.document.layers.selected,
+                nextLayers = nextState.document.layers.selected,
+                lastLayerProps = collection.pluckAll(lastLayers, ["id", "bounds"]),
+                nextLayerProps = collection.pluckAll(nextLayers, ["id", "bounds"]),
+                layersChanged = !Immutable.is(lastLayerProps, nextLayerProps);
+
+            return layersChanged;
+        },
+
+        componentWillMount: function () {
+            this._drawDebounced = _.debounce(this.drawOverlay, DEBOUNCE_DELAY);
+        },
+
+        componentWillUnmount: function () {
+            this._drawDebounced.cancel();
+        },
+
+        componentDidMount: function () {
+            this._drawDebounced();
+        },
+
+        componentDidUpdate: function () {
+            this._drawDebounced();
+        },
+
+        /**
+         * Calls all helper functions to draw super-select overlay
+         * Cleans it first
+         */
+        drawOverlay: function () {
+            var currentDocument = this.state.document,
+                svg = d3.select(ReactDOM.findDOMNode(this));
+
+            svg.selectAll(".artboard-adders").remove();
+            
+            if (!currentDocument || this.state.modalState) {
+                return null;
+            }
+            
+            if (this.state.vectorMaskMode) {
+                return null;
+            }
+
+            this._scrimGroup = svg.insert("g", ".transform-control-group")
+                .classed("artboard-adders", true)
+                .attr("transform", this.props.transformString);
+
+            // Reason we calculate the scale here is to make sure things like strokewidth / rotate area
+            // are not scaled with the SVG transform of the overlay
+            var transformObj = d3.transform(this._scrimGroup.attr("transform")),
+                layerTree = currentDocument.layers;
+
+            this._scale = 1 / transformObj.scale[0];
+            
+            this._drawArtboardAdders(svg, layerTree);
+        },
+
+        /**
+         * Calculates the new artboard bounds in the given direction
+         *
+         * @private
+         * @param {Bounds} bounds Artboard bounds to be copied
+         * @param {string} direction 
+         * @return {Bounds}
+         */
+        _getNewArtboardBounds: function (bounds, direction) {
+            var padding = 100,
```

Comment:
Padding is not related to screen drawing in this function, it's for checking to see if artboards we create on any side, with padding, will not intersect existing artboards.